### PR TITLE
Add `get_tablenames` method. Fix #71

### LINKED ISSUE #71: add method that lists all tables
would be very useful to know which tables exists. Could be a static function that takes a db file as argument

### PR DESCRIPTION
**Pull request by transfluxus**

### CODE REVIEW

File: sqlitedict.py [hunk 1]
```diff
@@ -105,6 +105,19 @@ def decode(obj):
     return loads(bytes(obj))
 
 
+def print_tablenames(filename):
+    """print tablenames and return them as list. Returns empty list if file does not exist (instead of creating that file)"""
+    if not os.path.isfile(filename):
+        logger.warning('file %s does not exist' %filename)
```
Comment (piskvorky) [inline sqlitedict.py]: Explicit fail (exception) better?

Also, spaces around `%`.

**Comment by ownport** [comment]
As proposal, instead of printing tables maybe it will be better to return the list of them and user will decide how to use this list?

File: sqlitedict.py [hunk 2]
```diff
@@ -105,6 +105,18 @@ def decode(obj):
...
 
 
+def print_tablenames(filename):
+    """print tablenames and return them as list. Returns empty list if file does not exist (instead of creating that file)"""
+    if not os.path.isfile(filename):
+        raise IOError('file %s does not exist' % (filename))
+    conn = sqlite3.connect(filename)
+    res = conn.execute("SELECT name FROM sqlite_master WHERE type='table';")
+    tablenames = [name[0] for name in res]
+    print(tablenames)
```
Comment (menshikh-iv) [inline sqlitedict.py]: `return tablenames` enough (no need to print it)

File: sqlitedict.py [hunk 3]
```diff
@@ -105,6 +105,18 @@ def decode(obj):
     return loads(bytes(obj))
 
 
+def print_tablenames(filename):
```
Comment (menshikh-iv) [inline sqlitedict.py]: IMO better to move it to SqliteDict class and use as method, wdyt @transfluxus ?

File: sqlitedict.py [hunk 4]
```diff
@@ -105,11 +105,10 @@ def decode(obj):
     return loads(bytes(obj))
 
 
-def print_tablenames(filename):
+def print_tablename(filename):
     """print tablenames and return them as list. Returns empty list if file does not exist (instead of creating that file)"""
     if not os.path.isfile(filename):
-        logger.warning('file %s does not exist' %filename)
-        return
+        raise IOError('file %s does not exist' % (filename))
```
[the remark this answers is not in the document]
Comment (piskvorky) [reply sqlitedict.py]: I see. That's really bad behaviour from sqlite3 -- I agree raising an exception is much better.

**Comment by transfluxus** [comment]
instead of this:
```
db = SqliteDict('testDB.sqlite')
db = SqliteDict('testDB.sqlite',tablename='vocab')
db.get_tablenames()
```

I would rather have something like this:
```
dbs = {tablename : SqliteDict('testDB.sqlite',tablename) for tablename in get_tablenames('testDB.sqlite')}
```

**Comment by transfluxus** [comment]
but just let me know, what you prefer. it includes both versions now

File: sqlitedict.py [hunk 5]
```diff
@@ -300,6 +299,12 @@ def clear(self):
         self.conn.execute(CLEAR_ALL)
         self.conn.commit()
 
+    def get_tablenames(self):
+        """print tablenames and return them as list"""
+        res = self.conn.select("SELECT name FROM sqlite_master WHERE type='table';")
```
Comment (piskvorky) [inline sqlitedict.py]: -1 on code duplication, it makes maintenance a headache. Keep it [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself).

Can you refactor the core functionality to live in only one place? Possibly being used/called from multiple places, that's fine.

**Comment by transfluxus** [comment]
Of course. That's what I came up with in the end, since the db connection/cursor are in the run of the multithread obj.

I'm not used to write test :), but I added some. Hope they make sense

File: .ipynb_checkpoints/Untitled-checkpoint.ipynb [hunk 6]
```diff
@@ -0,0 +1,6 @@
+{
```
Comment (menshikh-iv) [inline .ipynb_checkpoints/Untitled-checkpoint.ipynb]: unrelated file, please remove

File: tests/test_core.py [hunk 7]
```diff
@@ -262,3 +262,23 @@ def test_complex_struct(self):
...
+        db = sqlitedict.SqliteDict(fname) 
+        self.assertEqual(db.get_tablenames(), ['unnamed'])
+        db.close()
+
+        fname = norm_file('tests/db/tablenames-test-2.sqlite')
+        db1 = sqlitedict.SqliteDict(fname,tablename='table1')         
+        self.assertEqual(db1.get_tablenames(), ['table1'])
+        db2 = sqlitedict.SqliteDict(fname,tablename='table2') 
+        self.assertEqual(db1.get_tablenames(), ['table1','table2'])
+        db1.close()
```
Comment (menshikh-iv) [inline tests/test_core.py]:
```
with SliteDict(...): 
    self.assertEqual 
```
better than db = ; db.close()

File: tests/test_core.py [hunk 8]
```diff
@@ -262,3 +262,23 @@ def test_complex_struct(self):
         self.db['test'] = test_value
         assert self.db['test'] == test_value
         assert self.get_json('test') == json.dumps(test_value)
+
+
+class TablenamesTest(TestCaseBackport):
+
+    def test_tablenames(self):
+        fname = norm_file('tests/db/tablenames-test-1.sqlite')
+        db = sqlitedict.SqliteDict(fname)
```
Comment (menshikh-iv) [inline tests/test_core.py]: what's about an empty file?

File: sqlitedict.py [hunk 9]
```diff
@@ -105,6 +105,21 @@ def decode(obj):
     return loads(bytes(obj))
 
 
+def get_tablenames(filename):
```
Comment (menshikh-iv) [inline sqlitedict.py]: probably better to remove this function & move all code to method, wdyt?

File: .ipynb_checkpoints/Untitled-checkpoint.ipynb (hunk 6 above)
Comment (transfluxus) [reply .ipynb_checkpoints/Untitled-checkpoint.ipynb]: aii. sneaked in...

File: sqlitedict.py (hunk 9 above)
Comment (transfluxus) [reply sqlitedict.py]: you mean just a class method?
I can do that. just makes use case look unpractical. see above

File: tests/test_core.py (hunk 8 above)
Comment (transfluxus) [reply tests/test_core.py]: not sure what you mean. in any case table `unamed` will be created

File: tests/test_core.py (hunk 7 above)
Comment (transfluxus) [reply tests/test_core.py]: will do that

File: tests/test_core.py (hunk 8 above)
Comment (menshikh-iv) [reply tests/test_core.py]: @transfluxus aha, OK!

File: sqlitedict.py (hunk 9 above)
Comment (menshikh-iv) [reply sqlitedict.py]: @transfluxus what's about static method?

Comment (transfluxus) [reply sqlitedict.py]: ok

File: tests/test_core.py (hunk 7 above)
Comment (transfluxus) [reply tests/test_core.py]: is there a reason you're not having 
`from sqlitedict import SliteDict`  ?

Comment (menshikh-iv) [reply tests/test_core.py]: No

File: sqlitedict.py [hunk 10]
```diff
@@ -497,7 +510,7 @@ def select(self, req, arg=None):
                 break
             yield rec
 
-    def select_one(self, req, arg=None):
+    def select_one(self, req, arg=None):
```
Comment (menshikh-iv) [inline sqlitedict.py]: leading space (PEP8)

File: tests/test_temp_db.py [hunk 11]
```diff
@@ -11,7 +11,7 @@
 
 from sys import version_info
 major_version = version_info[0]
-
+from accessories import norm_file
```
Comment (menshikh-iv) [inline tests/test_temp_db.py]: unused import

File: sqlitedict.py [hunk 12]
```diff
@@ -288,6 +288,19 @@ def clear(self):
         self.conn.execute(CLEAR_ALL)
         self.conn.commit()
 
+    @staticmethod
+    def get_tablenames(filename):
+        """get the names of the tables in an sqlite db as a list"""
+        if not os.path.isfile(filename):
+            raise IOError('file %s does not exist' % (filename))   
+        conn = sqlite3.connect(filename)
```
Comment (menshikh-iv) [inline sqlitedict.py]: what's about `with` here too?

File: sqlitedict.py (hunk 10 above)
Comment (transfluxus) [reply sqlitedict.py]: is out

File: tests/test_temp_db.py (hunk 11 above)
Comment (transfluxus) [reply tests/test_temp_db.py]: is gone

File: sqlitedict.py (hunk 12 above)
Comment (transfluxus) [reply sqlitedict.py]: gotit

File: tests/test_core.py (hunk 7 above)
Comment (transfluxus) [reply tests/test_core.py]: ok, I did, makes things look cleaner.
having `with` makes things look better indeed. However my test is now the only one doing that. looks a bit inconsistent...

Comment (menshikh-iv) [reply tests/test_core.py]: @transfluxus I think we'll cleanup codebase later, don't worry :)

**Comment by menshikh-iv** [comment]
Thank you @transfluxus! Congratz with first contribution :1st_place_medal:
